src/javascript/url-shortener.js

```diff
@@ -12,71 +12,327 @@
+  // always available.
+  const bitlyClientId = window.BITLY_CLIENT_ID;
+  if (!bitlyClientId) {
+    throw new Error(
+      'No OAuth client ID available. Make sure to attach it as a ' +
+      'global in the server!'
+    );
+  }
+
+  // Initiate OAuth flow. Spawn a new window, which sends the user to bitly's
+  // auth page. That will redirect us to /url-shorten/auth-callback, which
+  // will get an authentication token, then send us a message via postMessage
+  // with the token.
+  //
+  // We can't use try-finally here because the various APIs we're using (
+  // addEventListener, timers, etc) aren't promisified or awaitable, so we have
+  // to wrap them up in a promise abstraction that cleans up after itself.
+  //
+  // TODO(nathanwest): RxJS was a relatively late addition to the URL shortener
+  // design, but now that we have it, see if any of this cleanup stuff can be
+  // more easily or cleanly expressed using that library.
+  const {token} = await cleanupingPromise((resolve, reject, cleanup) => {
+    // Attach a listener that will await a message via postMessage
+    const messageListener = event => {
+      if (
+        event.origin === window.location.origin &&
+        event.source === childWindow
+      ) {
+        const data = event.data;
+        if (data.error) {
+          reject(new Error(data.error));
+        } else {
+          resolve(data);
+        }
+      }
+    };
+    window.addEventListener('message', messageListener, {once: true});
+    cleanup(() => window.removeEventListener('message', messageListener));
+
+    const authUrl = 'https://bitly.com/oauth/authorize' + encodeQuery({
+      client_id: bitlyClientId,
+      // The auth callback does about half the work here: it receieves an
+      // authorization code, converts it into a token, then sends the token
+      // back to us. This happens server side because it requires the
+      // client_secret, which we don't want to share with our users.
+      redirect_uri: window.location.origin + '/bitly-api-token-handler/',
+    });
+    // Open a new window to do authentication.
+    const childWindow = window.open(authUrl, '_blank');
+    if (childWindow == null) {
+      throw new Error(
+        'Failed open new window for authorizing bitly. ' +
+        'Are you blocking popups?'
+      );
+    }
+
+    // Close the window as soon as auth flow completes.
+    cleanup(() => childWindow.close());
+
+    // If the user closes the window, that's an error
+    const intervalHandle = window.setInterval(() => {
+      if (childWindow.closed) {
+        reject(new Error(
+          'Authorization window closed before authorization was completed'
+        ));
+      }
+    }, 1000);
+    cleanup(() => window.clearInterval(intervalHandle));
+
+    // If the user takes too long, that's an error.
+    const timeoutHandle = window.setTimeout(() => {
+      reject(new Error('Timed out waiting for authorization'));
+    }, 1000 * 60 * 5);
```

Comment:
These would be better as named constants.

Reply:
Done